splunk-otel-android/src/test/java/com/splunk/rum/AnrWatcherTest.java

```diff
@@ -0,0 +1,101 @@
+ * limitations under the License.
+ */
+
+package com.splunk.rum;
+
+import android.os.Handler;
+
+import org.junit.Test;
+
+import static org.mockito.ArgumentMatchers.isA;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.when;
+
+public class AnrWatcherTest {
+
+    @Test
+    public void mainThreadDisappearing() {
+        Handler handler = mock(Handler.class);
+        Thread mainThread = mock(Thread.class);
+        SplunkRum splunkRum = mock(SplunkRum.class);
+
+        AnrWatcher anrWatcher = new AnrWatcher(handler, mainThread, () -> splunkRum);
+        for (int i = 0; i < 5; i++) {
+            when(handler.post(isA(Runnable.class))).thenReturn(false);
+            anrWatcher.run();
+        }
+        verifyNoInteractions(splunkRum);
+    }
+
+    @Test
+    public void noAnr() {
+        Handler handler = mock(Handler.class);
+        Thread mainThread = mock(Thread.class);
+        SplunkRum splunkRum = mock(SplunkRum.class);
+
+        AnrWatcher anrWatcher = new AnrWatcher(handler, mainThread, () -> splunkRum);
+        for (int i = 0; i < 5; i++) {
+            when(handler.post(isA(Runnable.class))).thenAnswer(invocation -> {
+                Runnable callback = (Runnable) invocation.getArgument(0);
+                callback.run();
+                return true;
+            });
+            anrWatcher.run();
+        }
+        verifyNoInteractions(splunkRum);
+    }
+
+    @Test
+    public void noAnr_temporaryPause() {
+        Handler handler = mock(Handler.class);
+        Thread mainThread = mock(Thread.class);
+        SplunkRum splunkRum = mock(SplunkRum.class);
+
+        AnrWatcher anrWatcher = new AnrWatcher(handler, mainThread, () -> splunkRum);
+        for (int i = 0; i < 5; i++) {
+            int index = i;
+            when(handler.post(isA(Runnable.class))).thenAnswer(invocation -> {
+                Runnable callback = (Runnable) invocation.getArgument(0);
+                //have it fail once
+                if (index != 3) {
+                    callback.run();
+                }
+                return true;
+            });
+            anrWatcher.run();
+        }
+        verifyNoInteractions(splunkRum);
+    }
+
+    @Test
+    public void anr_detected() {
+        Handler handler = mock(Handler.class);
+        Thread mainThread = mock(Thread.class);
+        SplunkRum splunkRum = mock(SplunkRum.class);
+
+        StackTraceElement[] stackTrace = new StackTraceElement[0];
+        when(mainThread.getStackTrace()).thenReturn(stackTrace);
+
+        AnrWatcher anrWatcher = new AnrWatcher(handler, mainThread, () -> splunkRum);
+        for (int i = 0; i < 5; i++) {
+            when(handler.post(isA(Runnable.class))).thenReturn(true);
+            anrWatcher.run();
+        }
+        verify(splunkRum).recordAnr(stackTrace);
+    }
```

Comment:
How about a test case that verifies that only one ANR is reported if the main thread does not respond for 6-9 seconds?

Reply:
Good idea. I'll add it.